Port 1.15.2 Villager optimizations

MrIvanPlays opened the pull request:
~~Tested & works without issues.
Also added some more obfuscation helpers for easier maintenance & update between major releases~~
raids don't work with it

Commit:
Port 1.15.2 Villager optimizations

Tested & works without issues.
Also added some more obfuscation helpers for easier maintenance & update between major releases

kennytv commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -174,31 +194,43 @@ index a6d8ef5eb44f3f851a3a1be4032ca21ab1d7f2b2..c3e8a0145d63843736d2060f978cdf38
      }
  
-+    public final boolean hasVacancy() { return d(); } // Paper - OBFHELPER
++    public final boolean hasVacany() { return d(); } // Paper - OBFHELPER
```
_ah yes, the vacany_

MrIvanPlays replied:
Whoops, typos 
Had to reimplement the whole patch by hand :eyes:

kennytv commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -121,51 +156,44 @@ index a6d8ef5eb44f3f851a3a1be4032ca21ab1d7f2b2..c3e8a0145d63843736d2060f978cdf38
-+                //this.c.put(SectionPosition.a(chunkcoordintpair, i).v(), Optional.empty()); // Paper - NO!!!
+                 this.c.put(SectionPosition.a(chunkcoordintpair, i).s(), Optional.empty());
              }
++             */ // Paper end
```
no

MrIvanPlays replied:
Why waste cpu time on even initializing the loop? 
Sure, as you wish

BillyGalbreath replied:
The compiler optimizes it out

Proximyst commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -93,25 +105,40 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
          }
--
-+        */ // Paper end
++         */ // Paper end
```
9 spaces?

MrIvanPlays replied:
Um wtf happened here? I remember intellij autocompleting that

Proximyst replied:
Intellisense doesn't apply to whitespace...

MrIvanPlays replied:
bruh
whatever now's been fixed

Proximyst commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -38,30 +38,42 @@ the code rebuilding data that is static every single invocation for every POI ty
 a plugin manipulates and adds custom professions, which it will handle by rebuilding.
 
+Ported to 1.16.1 by MrIvanPlays
+Co-authored-by: MrIvanPlays <ivan@mrivanplays.com>
```
It's cool to want attribution, but you didn't co-author this patch.

MrIvanPlays replied:
I want to see this commented from a papermc member otherwise it would stay.

JRoy replied:
Multiple people have ported patches over to new major MC versions countless times. They've never placed their name inside the patch for credit.

Your "port" is no different in that regard, git commit history is here for that.

mikroskeem replied:
Do not slap your name on things you haven't done. This is also the reason why you were banned from the IRC or Paper's Discord Guild.

duplexsystem replied:
I think his confusion comes from @Proximyst saying "If it was based on someone else's but you didn't write the base, it's still their patch. You'd write you co-authored it in that case." In response to Ivan saying "The Pathfinder changes don't exactly match aikar's changes in the previous patch."

MrIvanPlays replied:
pretty sure the reason behind why you don't want it there is because i'm banned from paper discord and you don't want to see me as a contributor. as i said i want to see this commented from papermc member (prefferably, the author of the patch - aikar) and then ill remove it.

> This is also the reason why you were banned from the IRC or Paper's Discord Guild.

It wasn't for that, no one minded I stole 1 patch of your microcord which later I removed. Now I did change and im trying to show it off by starting to contribute things and being more active in the overall minecraft discords, but the paper discord - no one wants to talk with me in other discords and i can't show it off to them.

> Multiple people have ported patches over to new major MC versions countless times. They've never placed their name inside the patch for credit.

valid point, but still want to see it commented from a papermc member.

pop4959 replied:
Why are you still arguing about this when it's clear that you're wrong? It doesn't make sense to wait for Aikar to tell you the exact same thing 5 people have already told you. I don't think anyone cares whether you are banned on the Discord or not, contributing is great, and everyone loves you for that. Just don't insist that you put your name on something that isn't your work.

BillyGalbreath replied:
> you don't want to see me as a contributor

Uh.. Even just with 1 commit you will end up in the list of contributors.. https://github.com/PaperMC/Paper/graphs/contributors

Best to handle things the right way, you're just making a clown out of yourself how you're doing it now.

MrIvanPlays replied:
> and everyone loves you for that

nice joke. i still think you're hating me no matter what. even when mini was porting paper to 1.16.1 and started a test server some1 just banned me for no reason. you all fucking hate me, you can stop trying to "show-off" that you actually care about the patch. you only care me to not be a contributor and to NOT participate in anything paper-related (and just for the record, this is NOT towards papermc members, cuz I know some of them do hate me but even then still talk with me, this is towards all of you who started the whole argument).

jacobsandersen replied:
Ivan, can you stop making a fool of yourself for 3 minutes?

You didn't author nor coauthor the patch, Aikar did. You updated it. They are asking you to remove your coauthor attribution because you are not the coauthor of the patch.

This is not in any way different from you choosing any book, updating it, and then slapping your name next to the author. It wouldn't be true and neither is this.

You're free to contribute. If you weren't, you'd be banned from even making issues/PRs so seriously, stop.

You claim all day and night that you've changed, but you literally have not. You love to spout off about how we're targeting you but you are doing specific actions repeatedly that do not align with Paper's goals. If you don't stop, then yeah you're going to get banned and nobody is going to like your attitude.

But that doesn't preclude you from still taking part of the community if you change your ways.

My 2c. Whatever. I doubt anyone cares so much about this patch that they are willing to do drama over it. 

TLDR remove the coauthor status or close the PR, but stop being so annoying.

MiniDigger replied:
Brb, adding myself as co-author to the ~200 patches I updated to 1.16

aikar replied:
There is nothing significant changed here to warrant a co author. So please remove.

Authorship is more about ideas and implementation design, not the syntax of a line.

Proximyst replied:
It was already removed with https://github.com/PaperMC/Paper/pull/4101/commits/ce978bbff0d937df7a966474779a4ea4c52bbabb

aikar commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -70,19 +79,19 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
 +
-+                            if (pathentity != null && pathentity.h()) {
++                            if (pathentity != null && pathentity.canReach()) {
 +                                record.decreaseVacancy();
```
This line has to be dropped. It's the only thing different in 1.16 (decreaseVacancy). They no longer claim the workstation upon targeting it, only when reaching it.

MrIvanPlays replied:
Done

aikar commented on Spigot-Server-Patches/0551-Optimize-Villagers.patch:
```diff
@@ -93,25 +102,41 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
++
+         this.f.long2ObjectEntrySet().removeIf((entry) -> {
+-            return !((BehaviorFindPosition.a) entry.getValue()).b(i);
++            return entry.getValue().b < e;
```
why was this changed?

MrIvanPlays replied:
To fit with your changes. See your original filtering (entry.getLongValue() is now entry.getValue().b)

MrIvanPlays commented on Spigot-Server-Patches/0174-revert-serverside-behavior-of-keepalives.patch:
```diff
@@ -17,7 +17,7 @@ from networking or during connections flood of chunk packets on slower clients,
 
 diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
-index 1d28b2b382877bb24958f28739e90bacfe8b9a94..76cd3ea9512f4355a055d9c683d939680dc5a1b7 100644
+index 985420238c9da53ab3b11d531f002a3d5f9a3d73..a1dd21a1b4ee6b3c6bbed545f75582730871ea98 100644
```
I have 0 idea of what happened here, pretty sure it's ok?

aikar reviewed (dismissed):
@PaperMC/development-team please spot check behavior of villages and claiming workstations before merge, but seems fine.

Proximyst commented on Spigot-Server-Patches/0565-Optimize-Villagers.patch:
```diff
@@ -70,19 +79,18 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
++                            PathEntity pathentity = entitycreature.getNavigation().a(com.google.common.collect.ImmutableSet.of(pos), 8, false, this.b.getValidRange());
 +
-+                            if (pathentity != null && pathentity.h()) {
-+                                record.decreaseVacancy();
```
Where'd the `decreaseVacancy` call go? We still need to make it clear that it's *not* available, do we not?

MrIvanPlays replied:
![image](https://img.mrivanplays.com/msedge_HYKypQGkUS.png)

Commit:
Fix raid issues

Thanks Proxi

Co-authored-by: Mariell Hoversholm <proximyst@proximyst.com>

Commit:
Merge branch 'master' of https://github.com/PaperMC/Paper into patch/port-villager-optimizations

Commit:
Rebuild patches

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -39,29 +39,47 @@ So we cache that and only rebuild it if professions change, which should be neve
 +        int requiredDist = dist * dist;
 +        int cdist = Math.floorDiv(dist, 16);
-+        Predicate<VillagePlaceType> predicate = this.a.c();
++        Predicate<VillagePlaceType> predicate = this.b.getPredicate();
```
`this.b` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -39,29 +39,47 @@ So we cache that and only rebuild it if professions change, which should be neve
 +        for (ChunkCoordIntPair chunkcoordintpair : MCUtil.getSpiralOutChunks(blockposition2, cdist)) {
 +            for (int i1 = 0; i1 < 16; i1++) {
-+                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).v());
++                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).asLong());
```
`SectionPosition.a` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -70,19 +88,19 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
 +                        BlockPosition pos = record.getPosition();
 +                        long key = pos.asLong();
-+                        if (this.e.containsKey(key)) {
++                        if (this.g.containsKey(key)) {
```
`this.g` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -70,19 +88,19 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
 +                        double poiDist = pos.distanceSquared(blockposition2);
 +                        if (poiDist <= (double) requiredDist) {
-+                            this.e.put(key, (long) (this.d + Math.sqrt(poiDist) * 4)); // use dist instead of 40 to blacklist longer if farther distance
++                            this.g.put(key, new BehaviorFindPosition.a(java.util.concurrent.ThreadLocalRandom.current(), (long) (this.f + Math.sqrt(poiDist) * 4))); // use dist instead of 40 to blacklist longer if farther distance
```
`this.g` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -70,19 +88,19 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
++                            this.g.put(key, new BehaviorFindPosition.a(java.util.concurrent.ThreadLocalRandom.current(), (long) (this.f + Math.sqrt(poiDist) * 4))); // use dist instead of 40 to blacklist longer if farther distance
 +                            ++poiAttempts;
-+                            PathEntity pathentity = entitycreature.getNavigation().a(com.google.common.collect.ImmutableSet.of(pos), 8, false, this.a.d());
++                            PathEntity pathentity = entitycreature.getNavigation().a(com.google.common.collect.ImmutableSet.of(pos), 8, false, this.b.getValidRange());
```
`this.b` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -70,19 +88,19 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
-+                                entitycreature.getBehaviorController().setMemory(this.b, globalPos);
++                            if (pathentity != null && pathentity.canReach()) {
++                                GlobalPos globalPos = GlobalPos.create(worldserver.getDimensionKey(), pos);
++                                entitycreature.getBehaviorController().setMemory(c, globalPos);
```
What is `c`?

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -39,29 +39,47 @@ So we cache that and only rebuild it if professions change, which should be neve
-+                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).v());
++                java.util.Optional<VillagePlaceSection> section = villageplace.getSection(SectionPosition.a(chunkcoordintpair, i1).asLong());
 +                if (section == null || !section.isPresent()) continue;
 +                for (java.util.Map.Entry<VillagePlaceType, java.util.Set<VillagePlaceRecord>> e : section.get().getRecords().entrySet()) {
```
Please use readable names; `e` tells us nothing, especially in the middle of obfuscated code.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -93,25 +111,40 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
++
+         this.g.long2ObjectEntrySet().removeIf((entry) -> {
+-            return !((BehaviorFindPosition.a) entry.getValue()).b(i);
++            return entry.getValue().b < f;
```
`b`? `f`?

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -93,25 +111,40 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
-+        // Clean up - vanilla does this only when it runs out, but that would push it to try farther POI's...
-+        this.e.long2LongEntrySet().removeIf((entry) -> entry.getLongValue() < this.d);
++
+         this.g.long2ObjectEntrySet().removeIf((entry) -> {
```
`this.g` is what? Obfhelper. I know this isn't modified, but it will help when updating the code.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -93,25 +111,40 @@ index 35eb3a5a61145e94d5b0c77c0eb13bfa46fac23b..6861b1a345496a83900b0ef702ba3431
 +        /*
          Predicate<BlockPosition> predicate = (blockposition) -> {
-             long j = blockposition.asLong();
+             BehaviorFindPosition.a behaviorfindposition_a = (BehaviorFindPosition.a) this.g.get(blockposition.asLong());
```
`this.g` is what? Obfhelper. I know this isn't modified, but it will help when updating the code.

MrIvanPlays replied:
This is mojang's original code. I doubt you need something changed here.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -121,51 +154,41 @@ index a6d8ef5eb44f3f851a3a1be4032ca21ab1d7f2b2..c3e8a0145d63843736d2060f978cdf38
+         }
++         */
 +        // If it's an unloaded chunk, well too bad.
 +        return this.c(i);
```
`this.c` is what? Obfhelper.

Proximyst commented on Spigot-Server-Patches/0576-Optimize-Villagers.patch:
```diff
@@ -121,51 +154,41 @@ index a6d8ef5eb44f3f851a3a1be4032ca21ab1d7f2b2..c3e8a0145d63843736d2060f978cdf38
                  });
  
 -                this.c.put(i1, optional);
 +                if (optional.isPresent()) this.c.put(i1, optional); // Paper - NO!!!
```
`this.c` is what? Obfhelper.

Commit:
Merge branch 'master' of https://github.com/PaperMC/Paper into patch/port-villager-optimizations

Commit:
Make it compile

Commit:
Amend with asked changes

Also renamed VillagePlaceType#getPredicate to getCompletionCondition, more descriptive

Commit:
Merge branch 'master' of https://github.com/MrIvanPlays/Paper into patch/port-villager-optimizations

Commit:
Rebuild patches
